internal/server/server.go

```diff
@@ -0,0 +1,296 @@
+	Address string
+	// Token is the token to use for authentication (optional)
+	Token string
+	// Templates is the list of templates to use for fuzzing
+	Templates []string
+	// Verbose is a flag that controls verbose output
+	Verbose bool
+
+	// Scope fields for fuzzer
+	InScope  []string
+	OutScope []string
+
+	OutputWriter output.Writer
+
+	NucleiExecutorOptions *NucleiExecutorOptions
+}
+
+// New creates a new instance of the DAST server.
+func New(options *Options) (*DASTServer, error) {
+	// If the user has specified no templates, use the default ones
+	// for DAST only.
+	if len(options.Templates) == 0 {
+		options.Templates = []string{"dast/"}
+	}
+	// Disable bulk mode and single threaded execution
+	// by auto adjusting in case of default values
+	if options.NucleiExecutorOptions.Options.BulkSize == 25 && options.NucleiExecutorOptions.Options.TemplateThreads == 25 {
+		options.NucleiExecutorOptions.Options.BulkSize = 1
+		options.NucleiExecutorOptions.Options.TemplateThreads = 1
+	}
+	maxWorkers := env.GetEnvOrDefault[int]("FUZZ_MAX_WORKERS", 1)
+	bufferSize := env.GetEnvOrDefault[int]("FUZZ_BUFFER_SIZE", 10000)
+
+	server := &DASTServer{
+		options:      options,
+		tasksPool:    pond.New(maxWorkers, bufferSize),
+		deduplicator: newRequestDeduplicator(),
+		startTime:    time.Now(),
+	}
+	server.setupHandlers(false)
+
+	executor, err := newNucleiExecutor(options.NucleiExecutorOptions)
+	if err != nil {
+		return nil, err
+	}
+	server.nucleiExecutor = executor
+
+	scopeManager, err := scope.NewManager(
+		options.InScope,
+		options.OutScope,
+	)
+	if err != nil {
+		return nil, err
+	}
+	server.scopeManager = scopeManager
+
+	var builder strings.Builder
+	builder.WriteString(fmt.Sprintf("Using %d parallel tasks with %d buffer", maxWorkers, bufferSize))
+	if options.Token != "" {
+		builder.WriteString(" (with token)")
+	}
+	gologger.Info().Msgf("%s", builder.String())
+	gologger.Info().Msgf("Connection URL: %s", server.buildConnectionURL())
+	gologger.Info().Msgf("Stats UI URL: %s", server.buildStatsURL())
+
+	return server, nil
+}
+
+func NewStatsServer(fuzzStatsDB *stats.Tracker) (*DASTServer, error) {
+	server := &DASTServer{
+		nucleiExecutor: &nucleiExecutor{
+			executorOpts: protocols.ExecutorOptions{
+				FuzzStatsDB: fuzzStatsDB,
+			},
+		},
+	}
+	server.setupHandlers(true)
+	gologger.Info().Msgf("Stats UI URL: %s", server.buildStatsURL())
+
+	return server, nil
+}
+
+func (s *DASTServer) Close() {
+	s.nucleiExecutor.Close()
+	s.echo.Close()
+	s.tasksPool.StopAndWaitFor(1 * time.Minute)
+}
+
+func (s *DASTServer) buildConnectionURL() string {
+	url := fmt.Sprintf("http://%s/requests", s.options.Address)
+	if s.options.Token != "" {
+		url += "?token=" + s.options.Token
+	}
+	return url
+}
+
+func (s *DASTServer) buildStatsURL() string {
+	url := fmt.Sprintf("http://%s/stats", s.options.Address)
+	if s.options.Token != "" {
+		url += "?token=" + s.options.Token
+	}
+	return url
+}
```

Comment:
This should use a URL builder; otherwise, it could lead to query parameter injection.